### PR TITLE
feature (new integration): Onboarding RockerBox Integration

### PR DESCRIPTION
**Pull request by yashasvibajpai**
## Description of the change

> Onboarding RockerBox Integration

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: constants/destinationCanonicalNames.js [hunk 1]
```diff
@@ -40,6 +40,7 @@ const DestCanonicalNames = {
   mailchimp: ["mailchimp", "MailChimp", "MAILCHIMP"],
   kafka: ["KAFKA", "kafka", "Kafka"],
   vero: ["vero", "Vero", "VERO"],
+  rockerbox: ["rockerbox", "ROCKERBOX", "Rockerbox", "RockerBox"],
```
Comment (ItsSudip) [inline constants/destinationCanonicalNames.js:43]:
```suggestion
  rockerbox: ["rockerbox", "ROCKERBOX", "Rockerbox", "RockerBox", "rockerBox"],
```

File: v0/destinations/rockerbox/config.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
+const { getMappingConfig } = require("../../util");
+
+const WEBHOOKBASEENDPOINT = "https://webhooks.getrockerbox.com/webhook/data";
```
Comment (ItsSudip) [inline v0/destinations/rockerbox/config.js:3]:
```suggestion
const BASE_ENDPOINT = "https://webhooks.getrockerbox.com/webhook/data";
```

File: v0/destinations/rockerbox/data/RockerboxTrackConfig.json [hunk 3]
```diff
@@ -0,0 +1,40 @@
+[
+  {
+    "sourceKeys": "userId",
+    "destKey": "customer_id",
+    "sourceFromGenericMap": true
+  },
+  {
+    "sourceKeys": ["context.traits.email", "properties.email"],
```
Comment (ItsSudip) [inline v0/destinations/rockerbox/data/RockerboxTrackConfig.json:8]:
```suggestion
    "sourceKeys": "email",
```

File: v0/destinations/rockerbox/data/RockerboxTrackConfig.json [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    "destKey": "email",
+    "sourceFromGenericMap": true
+  },
+  {
+    "sourceKeys": "context.traits.phone",
+    "destKey": "phone"
```
Comment (ItsSudip) [inline v0/destinations/rockerbox/data/RockerboxTrackConfig.json:14]:
```suggestion
    "sourceKeys": "phone",
    "destKey": "phone",
    "sourceFromGenericMap": true
```

File: v0/destinations/rockerbox/data/RockerboxTrackConfig.json [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  },
+  {
+    "sourceKeys": "originalTimestamp",
+    "destKey": "timestamp",
+    "metadata": {
+      "type": "secondTimestamp"
+    }
```
Comment (ItsSudip) [inline v0/destinations/rockerbox/data/RockerboxTrackConfig.json:21]:
```suggestion
    "sourceKeys": "timestamp",
    "destKey": "timestamp",
    "metadata": {
      "type": "secondTimestamp"
    },
    "sourceFromGenericMap": true
```

File: v0/destinations/rockerbox/data/RockerboxTrackConfig.json [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+  },
+
+  {
+    "sourceKeys": "properties.revenue",
```
Comment (ItsSudip) [inline v0/destinations/rockerbox/data/RockerboxTrackConfig.json:25]:
```suggestion
    "sourceKeys": ["properties.revenue", "properties.value", "properties.price", "properties.total"],
```

**Review by ItsSudip — CHANGES_REQUESTED** [review]

File: __tests__/data/rockerbox_router_input.json [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+        "customDomain": {
+          "web": "https://cookiedomain.com"
+        },
+        "enableCookieSync": {
+          "web": true
+        }
+      }
+    },
+    "metadata": {
+      "jobId": 1
```
Comment (utsabc) [inline __tests__/data/rockerbox_router_input.json:128]: Why jobId is 1 here as well it should be 2

File: v0/destinations/rockerbox/transform.js [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+  const { advertiserId, eventsMap } = destination.Config;
+  const eventsHashMap = getHashFromArray(eventsMap);
+
+  if (!eventsHashMap[message.event.toLowerCase()]) {
+    throw new CustomError(
+      "The event is not associated to a RockerBox event. Aborting!",
+      400
+    );
+  } else {
+    set((payload.action = eventsHashMap[message.event.toLowerCase()]));
```
Comment (utsabc) [inline v0/destinations/rockerbox/transform.js:30]:
```suggestion
    payload.action = eventsHashMap[message.event.toLowerCase()];
```
No need to use set here. Also the use was incorrect

**Review by utsabc — CHANGES_REQUESTED** [review]

File: __tests__/data/rockerbox_router_input.json (hunk 7 above)
Comment (yashasvibajpai) [reply __tests__/data/rockerbox_router_input.json:128]: Fixed.

File: v0/destinations/rockerbox/transform.js (hunk 8 above)
Comment (yashasvibajpai) [reply v0/destinations/rockerbox/transform.js]: fixed

File: constants/destinationCanonicalNames.js (hunk 1 above)
Comment (yashasvibajpai) [reply constants/destinationCanonicalNames.js]: addressed.

File: v0/destinations/rockerbox/config.js (hunk 2 above)
Comment (yashasvibajpai) [reply v0/destinations/rockerbox/config.js]: addressed.

File: v0/destinations/rockerbox/data/RockerboxTrackConfig.json (hunk 3 above)
Comment (yashasvibajpai) [reply v0/destinations/rockerbox/data/RockerboxTrackConfig.json]: addressed.

File: v0/destinations/rockerbox/data/RockerboxTrackConfig.json (hunk 4 above)
Comment (yashasvibajpai) [reply v0/destinations/rockerbox/data/RockerboxTrackConfig.json]: addressed.

File: v0/destinations/rockerbox/data/RockerboxTrackConfig.json (hunk 6 above)
Comment (yashasvibajpai) [reply v0/destinations/rockerbox/data/RockerboxTrackConfig.json]: addressed.

File: v0/destinations/rockerbox/transform.js [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+  return response;
+};
+
+const trackResponseBuilder = (message, category, destination) => {
+  const respList = [];
+  const response = responseBuilderSimple(message, category, destination);
+  respList.push(response);
+
+  return respList;
+};
```
[the remark this answers is not in the document]
Comment (yashasvibajpai) [reply v0/destinations/rockerbox/transform.js]: Removed the extra function.

File: v0/destinations/rockerbox/data/RockerboxTrackConfig.json [hunk 10]
```diff
@@ -5,24 +5,30 @@
     "sourceFromGenericMap": true
   },
   {
-    "sourceKeys": ["context.traits.email", "properties.email"],
+    "sourceKeys": ["email"],
```
Comment (utsabc) [inline v0/destinations/rockerbox/data/RockerboxTrackConfig.json:8]: No array required if its a single key (sourceKey)

Comment (yashasvibajpai) [reply v0/destinations/rockerbox/data/RockerboxTrackConfig.json]: fixed the same

**Review by utsabc — DISMISSED** [review]

File: v0/destinations/rockerbox/data/RockerboxTrackConfig.json [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    "destKey": "phone",
+    "sourceFromGenericMap": true
+  },
+  {
+    "sourceKeys": "timestamp",
+    "destKey": "timestamp",
+    "metadata": {
+      "type": "secondTimestamp"
+    },
+    "sourceFromGenericMap": true
```
Comment (ItsSudip) [inline v0/destinations/rockerbox/data/RockerboxTrackConfig.json:23]: https://help.rockerbox.com/docs/webhooks some fields are required, so make them required.

Comment (yashasvibajpai) [reply v0/destinations/rockerbox/data/RockerboxTrackConfig.json:26]: @ItsSudip Had missed the comment. Addressed.

**Review by ItsSudip — APPROVED** [review]

**Review by utsabc — APPROVED** [review]

**Pull request merged**